Fold more Unicode Quote types

megangardner opened the pull request:
#### 02f58b69e5c311fa6c1d5b1730c2ed28be3e1497
<pre>
Fold more Unicode Quote types
<a href="https://bugs.webkit.org/show_bug.cgi?id=245121">https://bugs.webkit.org/show_bug.cgi?id=245121</a>
&lt;rdar://99854364&gt;

Reviewed by Wenson Hsieh.

Fold all known unicode quote marks. This should be taken care of
in the ICU framework, (rdar://98798562), but for now, fold them manually.

* LayoutTests/fast/text/find-quotes.html:
* Source/WTF/wtf/unicode/CharacterNames.h:
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::foldQuoteMark):
(WebCore::foldQuoteMarks):

Canonical link: <a href="https://commits.webkit.org/254465@main">https://commits.webkit.org/254465@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2113bcb131a7a49fd762ddd2a89ed4fd0067dfdf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98446 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154758 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32195 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27753 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81487 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92918 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94773 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25567 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76061 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25501 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80428 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80408 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68476 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/80838 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29977 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/14473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/74628 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29705 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15442 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26310 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3131 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38396 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/77490 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31846 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34535 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17146 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

whsieh commented on LayoutTests/fast/text/find-quotes.html:
```diff
@@ -22,6 +22,30 @@
+        var doubleHighReversed9QuotationMark = String.fromCharCode(0x201F);
+        var doubleLow9QuotationMark = String.fromCharCode(0x201E);
+        var doubleLowReversedQuotationMark = String.fromCharCode(0x2E42);
+        var doublePrimeQuotationMark = String.fromCharCode(0x301E);
```
Do you need to add `doubleLowReversed9QuotationMark` here as well?

darinadler commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -1645,6 +1667,28 @@ String foldQuoteMarks(const String& stringToFold)
     result = makeStringByReplacingAll(result, leftSingleQuotationMark, '\'');
     result = makeStringByReplacingAll(result, leftLowSingleQuotationMark, '\'');
     result = makeStringByReplacingAll(result, rightDoubleQuotationMark, '"');
+    result = makeStringByReplacingAll(result, singleLow9QuotationMark, '\'');
```
Now that we have so many additional quote types, I don’t think we should use a separate function call for each. If you like I can help write a version that is more efficient.

megangardner committed:
Fold more Unicode Quote types
https://bugs.webkit.org/show_bug.cgi?id=245121
<rdar://99854364>

Reviewed by Wenson Hsieh.

Fold all known unicode quote marks. This should be taken care of
in the ICU framework, (rdar://98798562), but for now, fold them manually.

* LayoutTests/fast/text/find-quotes.html:
* Source/WTF/wtf/unicode/CharacterNames.h:
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::foldQuoteMark):
(WebCore::foldQuoteMarks):

Canonical link: https://commits.webkit.org/254465@main
